planner/core/exhaust_physical_plans.go

```diff
@@ -1940,16 +1940,18 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
 	joins := make([]PhysicalPlan, 0, 8)
 	canPushToTiFlash := p.canPushToCop(kv.TiFlash)
 	if p.ctx.GetSessionVars().IsMPPAllowed() && canPushToTiFlash {
-		if p.shouldUseMPPBCJ() {
-			mppJoins := p.tryToGetMppHashJoin(prop, true)
-			if (p.preferJoinType & preferBCJoin) > 0 {
-				return mppJoins, true, nil
-			}
-			joins = append(joins, mppJoins...)
-		} else {
-			mppJoins := p.tryToGetMppHashJoin(prop, false)
-			joins = append(joins, mppJoins...)
+		shuffleJoins := p.tryToGetMppHashJoin(prop, false)
+		if (p.preferJoinType&preferShuffleJoin) > 0 && len(shuffleJoins) > 0 { // has shuffle_join hints
+			return shuffleJoins, true, nil
 		}
+		bcastJoins := p.tryToGetMppHashJoin(prop, true)
+		if len(bcastJoins) > 0 &&
+			((p.preferJoinType&preferBCJoin) > 0 || // specified by bcast_join hints
+				p.shouldUseMPPBCJ()) { // specified by heuristic rules
+			return bcastJoins, true, nil
+		}
+		joins = append(joins, bcastJoins...)
+		joins = append(joins, shuffleJoins...)
```

Comment:
If there is no hint and `p.shouldUseMPPBCJ` is false, the old code only adds `shuffleJoins` but the new code adds both `bcastJoins` and `shuffleJoins`. Is that an expected change?